server/graphql/schema.js

```diff
@@ -53,10 +55,55 @@ const typeDefs = gql`
 
 const resolvers = {
   Query: {
-    // TODO: implement query resolvers
+    user: async (_, args) => {
+      return await User.findById(args.id);
+    },
+    messages: async (_, { chatroomId }) => {
+      return await Message.find({ location: chatroomId });
+    },
+    chatrooms: async () => {
+      return await Chatroom.find({}).populate('tags');
+    },
   },
   Mutation: {
-    // TODO: implement mutation resolvers
+    login: async (_, { email, password }, context) => {
+      const user = await User.findOne({ email });
+      if (!user || !await user.isCorrectPassword(password)) {
+        throw new Error('Invalid credentials');
```

Comment:
good wording here: `invalid credentials` does not reveal password or username information to a potential attacker